### PR TITLE
adin2111: OA support and fixes

### PR DESCRIPTION
**Pull request by spectrum70**
Started to work on this driver, to optimize and add missing features.

### CODE REVIEW

**Comment by jpmur** [comment]
LGTM

**Comment by spectrum70** [comment]
Add other 2 small patches.

**Comment by spectrum70** [comment]
Hi,

Add Open Alliance SPI support.

This is actually the most complex patch of the set, btw, i reused existing Microchip Open Alliance module for this.
Waiting for any Maintainer feedbacks, especially on the way i am zeroing gpio structure, to be able to control chip select by sw, since there may be better ways (see @@ -781,7 +916,7)

File: drivers/ethernet/Kconfig.adin2111 [hunk 1]
```diff
@@ -49,6 +49,24 @@ config ETH_ADIN2111_SPI_CFG0
...
 	  or Protection Mode (OPEN Alliance) on the SPI Host Interface.
 
+DT_ETH_ADIN2111_PATH := $(dt_nodelabel_path,adin2111)
+DT_ETH_ADIN2111_SPI_OA := $(dt_node_bool_prop,$(DT_ETH_ADIN2111_PATH),spi-oa)
+
+config ETH_ADIN2111_SPI_OA
+	bool "SPI_OA"
+	default $(DT_ETH_ADIN2111_SPI_OA)
+	help
+	  Must be set when Open Alliance SPI is configured by HW bootstrap deep-switches.
```
Comment (carlocaione) [inline drivers/ethernet/Kconfig.adin2111:59]: Why is this needed? Can't we just use `DT_INST_NODE_HAS_PROP` or equivalent in the code without going through one additional Kconfig symbol?

File: drivers/ethernet/Kconfig.adin2111 [hunk 2]
```diff
@@ -49,6 +49,24 @@ config ETH_ADIN2111_SPI_CFG0
...
+
+DT_ETH_ADIN2111_SPI_OA_PROTECTION := $(dt_node_bool_prop,$(DT_ETH_ADIN2111_PATH),spi-oa-protection)
+
+config ETH_ADIN2111_SPI_OA_PROTECTION
+	bool "SPI_OA_PROTECTION"
+	default $(DT_ETH_ADIN2111_SPI_OA_PROTECTION)
+	depends on $(DT_ETH_ADIN2111_SPI_OA)
+	help
+	  Must be set when Open Alliance SPI protection is configured by HW bootstrap deep-switches.
```
Comment (carlocaione) [inline drivers/ethernet/Kconfig.adin2111:68]: ditto

**Review by carlocaione — CHANGES_REQUESTED** [review]

**Comment by carlocaione** [comment]
> Waiting for any Maintainer feedbacks, especially on the way i am zeroing gpio structure, to be able to control chip select by sw, since there may be better ways (see @@ -781,7 +916,7)

@gmarull ^

File: drivers/ethernet/Kconfig.adin2111 (hunk 1 above)
Comment (spectrum70) [reply drivers/ethernet/Kconfig.adin2111:59]: This was an intermediate choice, since there is the CRC option (CFG0) actually in Kconfig. In this way all options are visible also in menuconfig.:

(2048) Stack size for a thread that processes ADIN IRQ
(2) Priority for internal incoming packet handler
(100) IP buffer timeout
[*] SPI_CFG0
[*] SPI_OA
[ ] SPI_OA_PROTECTION

In a next step i can set all the options in devicetree only, eventually.

Comment (carlocaione) [reply drivers/ethernet/Kconfig.adin2111:59]: This is totally redundant and while there could be overlapping between DT and Kconfig it is alway better to just put the hardware configuration in DT when possible.

Intermediate steps are only going to increase the workload for all the parts involved (you, reviewers, maintaners, etc...) so I'd just suggest to move all the options into the DT already in this PR.

Comment (GeorgeCGV) [reply drivers/ethernet/Kconfig.adin2111:59]: @spectrum70 my mistake was to place the CRC option into Kconfig.

Ideally, the CRC should be moved into the DT. Some adin eval boards were recently added; and there it might be necessary to control CRC on/off at runtime (_due to board config switches_).

File: drivers/ethernet/eth_adin2111.c [hunk 3]
```diff
@@ -779,7 +916,7 @@ static int adin2111_await_device(const struct device *dev)
 
 static int adin2111_init(const struct device *dev)
 {
-	const struct adin2111_config *cfg = dev->config;
+	struct adin2111_config *cfg = (struct adin2111_config *)dev->config;
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:919]: What is the reason for `const` to disappear?

File: drivers/ethernet/eth_adin2111.c [hunk 4]
```diff
@@ -882,11 +1052,7 @@ static int adin2111_init(const struct device *dev)
 		return ret;
 	}
 
-#if CONFIG_ETH_ADIN2111_SPI_CFG0
 	val |= ADIN2111_CONFIG2_CRC_APPEND;
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:1055]: The CRC is now always enabled.

File: drivers/ethernet/eth_adin2111.c [hunk 5]
```diff
@@ -758,10 +891,14 @@ static int adin2111_await_device(const struct device *dev)
 
 	/* await reset complete (RESETC) and clear it */
 	for (count = 0U; count < ADIN2111_RESETC_AWAIT_RETRY_COUNT; ++count) {
-		ret = eth_adin2111_reg_read(dev, ADIN2111_STATUS0, &val);
+		ret = eth_adin2111_reg_read(dev, ADIN2111_PHYID, &val);
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:894]: Not sure about that change. The RM mentions STATUS0 poll. Will need to test.

Do you use HW reset? If no, then it could be related to clock stabilization delay (_done within hw reset_).

File: drivers/ethernet/eth_adin2111.c [hunk 6]
```diff
@@ -805,6 +942,25 @@ static int adin2111_init(const struct device *dev)
...
 	}
 
+	if (IS_ENABLED(CONFIG_ETH_ADIN2111_SPI_OA)) {
+		struct gpio_dt_spec spec = {0, 0, 0};
+
+		/*
+		 * OA requires sw chip select control
+		 */
+		ctx->cs_gpio = cfg->spi.config.cs.gpio;
+		cfg->spi.config.cs.gpio = spec;
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:952]: That is a bit odd as driver config is supposed to be immutable.

It would be better to solve during config definition with macros: `COND_CODE_1` and `DT_INST_NODE_HAS_PROP`.

When OA property is present, then `spi.config.cs.gpio` can be set to `{0, 0, 0}`, otherwise `SPI_DT_SPEC_INST_GET` is used.

That will require OA to be a DT property as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/68532/files#r1482973313.

File: drivers/ethernet/eth_adin2111_priv.h [hunk 7]
```diff
@@ -92,7 +101,7 @@
 /*!< Mask Bit for P1_RX_RDY. Generic SPI only.*/
 #define ADIN2111_IMASK1_P1_RX_RDY_MASK		BIT(4)
 /*!< Mask Bit for TX_FRM_DONE. Generic SPI only.*/
-#define ADIN2111_IMASK1_TX_RDY_MASK		BIT(4)
+#define ADIN2111_IMASK1_TX_RDY_MASK		BIT(3)
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111_priv.h:104]: Great catch.

File: drivers/ethernet/Kconfig.adin2111 (hunk 1 above)
Comment (spectrum70) [reply drivers/ethernet/Kconfig.adin2111:59]: Ok i proceed as required.

File: drivers/ethernet/eth_adin2111.c (hunk 4 above)
Comment (spectrum70) [reply drivers/ethernet/eth_adin2111.c:1055]: This is related to eth frame crc32, not connected to spi crc8

File: drivers/ethernet/eth_adin2111.c (hunk 5 above)
Comment (spectrum70) [reply drivers/ethernet/eth_adin2111.c:894]: Thaks, 

i tested mainly using the SW reset. Anyway i see that keeping the reset button pressed adin2111 clock stops.

Reading a read-only reg currently may not be the best, sure, but it works reliably, and allows "Generic SPI without CRC" to work.

What about if i move the check in the phy code ?

Anyway, asking also suggestions to the adi team tight now.



I see that

File: drivers/ethernet/eth_adin2111.c (hunk 6 above)
Comment (spectrum70) [reply drivers/ethernet/eth_adin2111.c:952]: Thanks for the feedback, will do as suggestd.

File: drivers/ethernet/eth_adin2111.c (hunk 5 above)
Comment (GeorgeCGV) [reply drivers/ethernet/eth_adin2111.c:894]: @spectrum70 in cases when hw reset is not used does increasing the delay from 10ms to `ADIN2111_HW_BOOT_DELAY_MS` helps?

File: drivers/ethernet/eth_adin2111.c (hunk 4 above)
Comment (GeorgeCGV) [reply drivers/ethernet/eth_adin2111.c:1055]: Rechecked RM, indeed.

**Comment by GeorgeCGV** [comment]
@spectrum70 thank you for the addition, cleanup, and fixes. Tested legacy spi with crc, it works. I am ok with a merge as soon as the driver config is kept immutable (_cs pin is set during config definition_).

**Comment by spectrum70** [comment]
In this PR:
- remove CONFIG_ stuff for Open Alliance, used DT_PROP and COND_CODE_1
- add oa and oa_protection to use/call Open Alliance features.

**Comment by spectrum70** [comment]
> @spectrum70 thank you for the addition, cleanup, and fixes. Tested legacy spi with crc, it works. I am ok with a merge as soon as the driver config is kept immutable (_cs pin is set during config definition_).

Thanks to you, a pleasure to work on this driver, nice and clean code.

**Review by carlocaione — DISMISSED** [review]

File: drivers/ethernet/eth_adin2111.c [hunk 8]
```diff
@@ -781,7 +915,7 @@ static int adin2111_await_device(const struct device *dev)
 
 static int adin2111_init(const struct device *dev)
 {
-	const struct adin2111_config *cfg = dev->config;
+	struct adin2111_config *cfg = (struct adin2111_config *)dev->config;
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:918]: please, add const back and remove the unnecessary cast

File: drivers/ethernet/eth_adin2111.c [hunk 9]
```diff
@@ -852,6 +1000,11 @@ static int adin2111_init(const struct device *dev)
 		return ret;
 	}
 
+	if (ctx->oa) {
+		/* OA mode, device needs some time to respond */
+		k_msleep(1);
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:1005]: "some time" feels too vague here (magic sleep), isn't there a minimum value specified in the datasheet?

File: drivers/ethernet/eth_adin2111.c [hunk 10]
```diff
@@ -839,6 +982,11 @@ static int adin2111_init(const struct device *dev)
 		return ret;
 	}
 
+	if (ctx->oa) {
+		/* OA mode, device needs some time to respond */
+		k_msleep(10);
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:987]: same here, magic value

File: drivers/ethernet/eth_adin2111.c [hunk 11]
```diff
@@ -966,21 +1133,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+						    SPI_MODE_CPHA, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_hold_cs),		\
+						    SPI_HOLD_ON_CS, (0)),			\
+				.slave = DT_REG_ADDR(DT_DRV_INST(inst)),			\
+				.cs = {								\
+					.gpio = COND_CODE_1(DT_INST_PROP_OR(inst, spi_oa, 0),	\
+						({0, 0, 0}),					\
+						(SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)))),	\
+					.delay = 0,						\
+				}								\
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:1159]: aren't SPI DT spec macros sufficient here?

File: drivers/ethernet/eth_adin2111.c [hunk 12]
```diff
@@ -966,21 +1133,49 @@ static const struct ethernet_api adin2111_port_api = {
...
 #define ADIN2111_SPI_OPERATION ((uint16_t)(SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | SPI_WORD_SET(8)))
-
 #define ADIN2111_MAC_INITIALIZE(inst, dev_id, ifaces, name)					\
 	static uint8_t __aligned(4) name##_buffer_##inst[CONFIG_ETH_ADIN2111_BUFFER_SIZE];	\
 	static const struct adin2111_config name##_config_##inst = {				\
-		.id = dev_id,								\
-		.spi = SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION, 1),			\
+		.id = dev_id,									\
+		.spi = {									\
+			.bus = DEVICE_DT_GET(DT_BUS(DT_DRV_INST(inst))),			\
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:1141]: please don't use DT_DRV_INST() when `DT_INST_*` macros exist

File: drivers/ethernet/eth_adin2111.c [hunk 13]
```diff
@@ -966,21 +1133,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+						(SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)))),	\
+					.delay = 0,						\
+				}								\
+			}									\
+		},										\
 		.interrupt = GPIO_DT_SPEC_INST_GET(inst, int_gpios),				\
 		.reset = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, { 0 }),			\
 	};											\
+	static struct oa_tc6 name##_oa_tc6_##inst = {						\
+		.cps = 64,									\
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:1166]: is this used? can't struct be const?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_adin2111.c [hunk 14]
```diff
@@ -968,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+						    SPI_MODE_CPOL, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpha),		\
+						    SPI_MODE_CPHA, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_hold_cs),		\
+						    SPI_HOLD_ON_CS, (0)),			\
+				.slave = DT_REG_ADDR(DT_DRV_INST(inst)),			\
+				.cs = {								\
+					.gpio = COND_CODE_1(DT_INST_PROP_OR(inst, spi_oa, 0),	\
+						({0, 0, 0}),					\
+						(SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)))),	\
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:1161]: if the whole objective of `spi-oa` is to override CS control, why not enforce it via a build assert, eg

```
BUILD_ASSERT(DT_INST_HAS_PROP(inst, spi_oa) && !DT_PROP_HAS_IDX(DT_INST_BUS(inst), DT_INST_REG_ADDR(inst));
```
then simply initialize SPI with `SPI_DT_SPEC_INST_GET`.

Another option would be to modify SPI API with a flag to override CS control even if non-NULL.

File: drivers/ethernet/eth_adin2111.c [hunk 15]
```diff
@@ -966,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+	};											\
 	static struct adin2111_data name##_data_##inst = {					\
 		.ifaces_left_to_init = ifaces,							\
 		.port = {},									\
 		.offload_sem = Z_SEM_INITIALIZER(name##_data_##inst.offload_sem, 0, 1),         \
 		.lock = Z_MUTEX_INITIALIZER(name##_data_##inst.lock),				\
 		.buf = name##_buffer_##inst,							\
+		.oa = DT_INST_PROP_OR(inst, spi_oa, 0),						\
+		.tc6 = &name##_oa_tc6_##inst,							\
+		.cs_gpio = SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst))				\
```
Comment (carlocaione) [inline drivers/ethernet/eth_adin2111.c:1182]: use `SPI_CS_GPIOS_DT_SPEC_INST_GET`. 

This is a general comment really. When you have something like `DT_*(DT_DRV_INST` in 90% of cases you have some `_INST_` instance of the same macro that let you omit `DT_DRV_INST`

File: drivers/ethernet/eth_adin2111.c (hunk 14 above)
Comment (gmarull) [reply drivers/ethernet/eth_adin2111.c:1161]: ping here

**Comment by spectrum70** [comment]
Forgot out am additional missing SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)) ,
changed to SPI_CS_GPIOS_DT_SPEC_INST_GET

**Comment by spectrum70** [comment]
I have seen that Open Alliance works reliably also with CS deasserted and asserted after each data block (OA chunk), so in case i can remove software control from the driver.

File: drivers/ethernet/eth_adin2111.c [hunk 16]
```diff
@@ -837,6 +981,12 @@ static int adin2111_init(const struct device *dev)
 		return ret;
 	}
 
+	/*
+	 * Setting this for all modes, RM rev.0 Table 3, t3 >= 50ms
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:985]:
```suggestion
	 * Setting this for all modes, RM rev. A Table 3, t3 >= 50ms
```

File: drivers/ethernet/eth_adin2111.c [hunk 17]
```diff
@@ -968,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+					DT_INST_PROP(inst, frame_format) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpol),		\
+						    SPI_MODE_CPOL, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpha),		\
+						    SPI_MODE_CPHA, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_hold_cs),		\
+						    SPI_HOLD_ON_CS, (0)),			\
+				.slave = DT_REG_ADDR(DT_DRV_INST(inst)),			\
+				.cs = {								\
+					.gpio = COND_CODE_1(DT_INST_PROP_OR(inst, spi_oa, 0),	\
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:1159]:
```suggestion
					.gpio = COND_CODE_1(DT_INST_PROP(inst, spi_oa),		\
```

File: drivers/ethernet/eth_adin2111.c (hunk 13 above)
Comment (GeorgeCGV) [reply drivers/ethernet/eth_adin2111.c:1170]: Struct internals are modified within `oa_tc6.c`, it seems to store config and data.

File: drivers/ethernet/eth_adin2111.c [hunk 18]
```diff
@@ -968,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+				}								\
+			}									\
+		},										\
 		.interrupt = GPIO_DT_SPEC_INST_GET(inst, int_gpios),				\
 		.reset = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, { 0 }),			\
 	};											\
+	static struct oa_tc6 name##_oa_tc6_##inst = {						\
+		.cps = 64,									\
+		.spi = &name##_config_##inst.spi,						\
+		.protected = DT_INST_PROP_OR(inst, spi_oa_protection, 0),			\
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:1172]:
```suggestion
		.protected = DT_INST_PROP(inst, spi_oa_protection),				\
```

**Comment by GeorgeCGV** [comment]
@spectrum70 regarding CS control and current approach. Won't added `SPI_LOCK_ON` to `ADIN2111_SPI_OPERATION` do the job? The SPI could be locked for the duration of OA calls. 

That way there is no need to define extra CS pin and duplicate `SPI_CONFIG_DT`. Then the `.spi` becomes:
```
		.spi = COND_CODE_1(DT_INST_PROP(inst, spi_oa),					\
			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION | SPI_LOCK_ON, 0),	\
			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION, 0)),			\
```

File: drivers/ethernet/eth_adin2111.c [hunk 19]
```diff
@@ -966,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
...
+			.config = {								\
+				.frequency = DT_INST_PROP(inst, spi_max_frequency),		\
+				.operation = (ADIN2111_SPI_OPERATION) |				\
+					DT_INST_PROP(inst, duplex) |				\
+					DT_INST_PROP(inst, frame_format) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpol),		\
+						    SPI_MODE_CPOL, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpha),		\
+						    SPI_MODE_CPHA, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_hold_cs),		\
```
[the remark this answers is not in the document]
Comment (GeorgeCGV) [reply drivers/ethernet/eth_adin2111.c:1155]: The `DT_INST_PROP_OR` won't work here. The expected output is a number (_some bit_); not plain `1` or `0`. It basically follows `SPI_CONFIG_DT` from `spi.h`.

File: drivers/ethernet/eth_adin2111.c (hunk 16 above)
Comment (GeorgeCGV) [reply drivers/ethernet/eth_adin2111.c:985]: [Publicly available RM](https://www.analog.com/media/en/technical-documentation/data-sheets/adin2111.pdf) is revision A.

**Comment by GeorgeCGV** [comment]
@spectrum70 considering other fixes, could you please add a minor correction to `k_thread_create` so it uses `K_KERNEL_STACK_SIZEOF` instead of `CONFIG_...`:

```
		k_thread_create(&ctx->rx_thread, ctx->rx_thread_stack,
				K_KERNEL_STACK_SIZEOF(ctx->rx_thread_stack),
				adin2111_offload_thread,
				(void *)adin, NULL, NULL,
				CONFIG_ETH_ADIN2111_IRQ_THREAD_PRIO,
				K_ESSENTIAL, K_NO_WAIT);
```

**Comment by spectrum70** [comment]
> @spectrum70 considering other fixes, could you please add a minor correction to `k_thread_create` so it uses `K_KERNEL_STACK_SIZEOF` instead of `CONFIG_...`:
> 
> ```
> 		k_thread_create(&ctx->rx_thread, ctx->rx_thread_stack,
> 				K_KERNEL_STACK_SIZEOF(ctx->rx_thread_stack),
> 				adin2111_offload_thread,
> 				(void *)adin, NULL, NULL,
> 				CONFIG_ETH_ADIN2111_IRQ_THREAD_PRIO,
> 				K_ESSENTIAL, K_NO_WAIT);
> ```

done

**Comment by spectrum70** [comment]
> @spectrum70 regarding CS control and current approach. Won't added `SPI_LOCK_ON` and `SPI_HOLD_ON_CS` (_maybe_) to `ADIN2111_SPI_OPERATION` do the job? The SPI could be locked for the duration of OA calls.
> 
> That way there is no need to define extra CS pin and duplicate `SPI_CONFIG_DT`. Then the `.spi` becomes:
> 
> ```
> 		.spi = COND_CODE_1(DT_INST_PROP(inst, spi_oa),					\
> 			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION | SPI_LOCK_ON, 0),	\
> 			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION, 0)),			\
> ```

It's ok, it works this way, thanks, so changed this way.

**Review by carlocaione — DISMISSED** [review]

**Comment by spectrum70** [comment]
@GeorgeCGV there was server-side issues with automated checks yesterday, now they are all passed.

File: drivers/ethernet/eth_adin2111.c (hunk 5 above)
Comment (GeorgeCGV) [reply drivers/ethernet/eth_adin2111.c:909]: > Anyway, i am asking also suggestions to the adi team right now.

Was there any response from them?

**Review by GeorgeCGV — DISMISSED** [review]
@spectrum70 thank you for the fixes and OA support, LGTM.

File: drivers/ethernet/eth_adin2111.c [hunk 20]
```diff
@@ -966,21 +1134,30 @@ static const struct ethernet_api adin2111_port_api = {
...
+		.spi = &name##_config_##inst.spi,						\
+		.protected = DT_INST_PROP(inst, spi_oa_protection),				\
+	};											\
 	static struct adin2111_data name##_data_##inst = {					\
 		.ifaces_left_to_init = ifaces,							\
 		.port = {},									\
 		.offload_sem = Z_SEM_INITIALIZER(name##_data_##inst.offload_sem, 0, 1),         \
 		.lock = Z_MUTEX_INITIALIZER(name##_data_##inst.lock),				\
 		.buf = name##_buffer_##inst,							\
+		.oa = DT_INST_PROP_OR(inst, spi_oa, 0),						\
```
Comment (gmarull) [inline drivers/ethernet/eth_adin2111.c:1159]: `DT_INST_PROP(inst, spi_oa)` is sufficient, since it's a boolean prop

**Review by GeorgeCGV — APPROVED** [review]

**Review by gmarull — DISMISSED** [review]

**Review by GeorgeCGV — APPROVED** [review]

**Review by GeorgeCGV — DISMISSED** [review]

**Comment by spectrum70** [comment]
Since there is still time before the next release, i added a further related patch on this branch enabling DMA, this way the ADI boards can reach 20Mbit/s of SPI transfer rate out of the box, needed for the maximum adin2111/1110 throughput.

**Review by carlocaione — DISMISSED** [review]

**Review by GeorgeCGV — DISMISSED** [review]

**Comment by spectrum70** [comment]
unfortunately, on further testing by zperf found  some OA related issues, freezing to solve issues.

**Comment by GeorgeCGV** [comment]
> stm32 re-transmit same tcp sequence packet

Does `eth_adin2111_oa_data_write` return any error? What is the state of `TXBOE`, `RXBOE`, and `TXBUE` in the `STATUS0` register when it happens (maybe modification of FIFO sizes could help if the host is not fast enough)?

File: drivers/ethernet/eth_adin2111.c [hunk 21]
```diff
@@ -149,13 +126,159 @@ int eth_adin2111_reg_read(const struct device *dev, const uint16_t reg,
...
+	int ret;
+
+	reg_oa = MMS_REG(reg >= ADIN2111_OA_MMS_MAC_OFFSET, reg);
+
+	ret = oa_tc6_reg_read(ctx->tc6, reg_oa, val);
+
+	spi_release_dt(&cfg->spi);
+
+	return ret;
+
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:180]:
```suggestion
```

File: drivers/ethernet/eth_adin2111.c [hunk 22]
```diff
@@ -418,6 +558,24 @@ static int adin2111_port_send(const struct device *dev, struct net_pkt *pkt)
...
+		 * (60 + OA header). Smaller packets are not sent out from the mac.
+		 */
+		if (pkt_len < ADIN2111_OA_MIN_PACKET_SIZE) {
+			net_buf_add(net_buf_frag_last(pkt->frags),
+				    ADIN2111_OA_MIN_PACKET_SIZE - pkt_len);
+		}
+
+		ret = eth_adin2111_oa_data_write(cfg->adin, pkt);
+		if (ret < 0) {
+			LOG_ERR("Failed to write OA spi data, %d", ret);
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:573]:
```suggestion
			eth_stats_update_errors_tx(data->iface);
			LOG_ERR("Failed to write OA spi data, %d", ret);
```

**Comment by spectrum70** [comment]
@GeorgeCGV when iperf stucks, mainly, the adin2111 receiving thread is blocked.
And loks like it is blocked at 

`/* await INT */
k_sem_take(&ctx->offload_sem, K_FOREVER);`

In this state, packets can still be sent out but nothing is received, becouse there is no hw interrupt
issued from adin2111 (checked by oscilloscope).

So strangely, only for Open Alliance, fast packet transmission by iperf locks the adin2111.
Still investigating.

**Comment by GeorgeCGV** [comment]
> Well, looks like i miss to check for space before sending a frame in OA mode.

That is great. Feel free to change the status to "Ready for review" when it works as expected.

**Comment by spectrum70** [comment]
Hi,
over hard zperf/iperf testing there was a couple of issues on using oa_tc6 developed for Microchip that i could not solve without putting hands in the module, that i cannot re-test for Microchip then. So i re-implemented Open Alliance part inside the driver, featuring a single dma transfer per frame, that works much better.

oa_tc6 common code cannot be used in a profitable way, due to : 
- not addressing port 2
- sends only chunk by chunk, in one dma transfer per chunk, while adin2111 allows a single dma transfer with all chiunks.

I am attaching some zperf tests.

[test-zephyr-adin2111ebz-iperf.txt](https://github.com/zephyrproject-rtos/zephyr/files/14483403/test-zephyr-adin2111ebz-iperf.txt)


Waiting for comments.

PS: there is still a brief outstanding issue open, not related to this job, but i am investigating on it. I noticed that on incoming udp packet of size 1512, generic spi fails on  "Still some length to go 2" net pkt message.

**Comment by spectrum70** [comment]
Fixed the issue reported above ("Still some length to go 2" net pkt message.).
Retested all, attaching zperf test.
[test-zephyr-adin2111ebz-iperf.txt](https://github.com/zephyrproject-rtos/zephyr/files/14493924/test-zephyr-adin2111ebz-iperf.txt)

**Comment by spectrum70** [comment]
Adding also new zperf board to board measurement, udp is at 8.5Mbit/s
[test-zephyr-adin2111ebz-iperf.txt](https://github.com/zephyrproject-rtos/zephyr/files/14512875/test-zephyr-adin2111ebz-iperf.txt)

**Review by carlocaione — APPROVED** [review]
not my usual cup of tea but LGTM for the general part

**Review by MaureenHelm — APPROVED** [review]

**Review by GeorgeCGV — APPROVED** [review]
Nice work, tested generic spi with enabled crc, everything works.

**Review by decsny — APPROVED** [review]

File: drivers/ethernet/eth_adin2111.c [hunk 23]
```diff
@@ -320,6 +637,22 @@ static void adin2111_offload_thread(void *p1, void *p2, void *p3)
...
+				}
+			}
+			if (status1 & ADIN2111_STATUS1_P2_RX_RDY) {
+				ret = eth_adin2111_oa_data_read(dev, 1);
+				if (ret < 0) {
+					break;
+				}
+			}
+			goto continue_unlock;
+		}
```
Comment (GeorgeCGV) [inline drivers/ethernet/eth_adin2111.c:654]: @spectrum70 , completely missed during the review, there should be no `break` for the `ctx->oa` code; as there is no `while`. The `break` will terminate the irq offload thread. Because the thread is `K_ESSENTIAL` the kernel should crash.
